server/src/main/java/org/opensearch/index/RemoteRefreshSegmentPressureService.java

```diff
@@ -0,0 +1,199 @@
+
+import java.util.Locale;
+import java.util.Map;
+import java.util.function.BiConsumer;
+
+/**
+ * Service used to validate if the incoming indexing request should be rejected based on the {@link RemoteRefreshSegmentPressureTracker}.
+ */
+public class RemoteRefreshSegmentPressureService implements IndexEventListener {
+
+    private static final Logger logger = LogManager.getLogger(RemoteRefreshSegmentPressureService.class);
+
+    /**
+     * Keeps map of remote-backed index shards and their corresponding backpressure tracker.
+     */
+    private final Map<ShardId, RemoteRefreshSegmentPressureTracker> trackerMap = ConcurrentCollections.newConcurrentMap();
+
+    /**
+     * Remote refresh segment pressure settings which is used for creation of the backpressure tracker and as well as rejection.
+     */
+    private final RemoteRefreshSegmentPressureSettings pressureSettings;
+
+    @Inject
+    public RemoteRefreshSegmentPressureService(ClusterService clusterService, Settings settings) {
+        pressureSettings = new RemoteRefreshSegmentPressureSettings(clusterService, settings, this);
+    }
+
+    /**
+     * Get {@code RemoteRefreshSegmentPressureTracker} only if the underlying Index has remote segments integration enabled.
+     *
+     * @param shardId shard id
+     * @return the tracker if index is remote-backed, else null.
+     */
+    public RemoteRefreshSegmentPressureTracker getPressureTracker(ShardId shardId) {
+        return trackerMap.get(shardId);
+    }
+
+    @Override
+    public void afterIndexShardCreated(IndexShard indexShard) {
+        if (indexShard.indexSettings().isRemoteStoreEnabled() == false) {
+            return;
+        }
+        ShardId shardId = indexShard.shardId();
+        trackerMap.put(shardId, new RemoteRefreshSegmentPressureTracker(shardId, pressureSettings));
+        logger.trace("Created tracker for shardId={}", shardId);
+    }
+
+    @Override
+    public void beforeIndexShardClosed(ShardId shardId, IndexShard indexShard, Settings indexSettings) {
+        if (indexShard.indexSettings().isRemoteStoreEnabled() == false) {
+            return;
+        }
+        trackerMap.remove(shardId);
+        logger.trace("Deleted tracker for shardId={}", shardId);
+    }
+
+    /**
+     * Check if remote refresh segments backpressure is enabled. This is backed by a cluster level setting.
+     *
+     * @return true if enabled, else false.
+     */
+    public boolean isSegmentsUploadBackpressureEnabled() {
+        return pressureSettings.isRemoteRefreshSegmentPressureEnabled();
+    }
+
+    public void validateSegmentsUploadLag(ShardId shardId) {
+        RemoteRefreshSegmentPressureTracker pressureTracker = getPressureTracker(shardId);
+        // Check if refresh checkpoint (a.k.a. seq number) lag is 2 or below - this is to handle segment merges that can
+        // increase the bytes to upload almost suddenly.
+        if (pressureTracker.getSeqNoLag() <= 2) {
```

Comment:
doubt: how is `2` decided here? in the test runs during segment merges how far this lag go? and how often it remained <= 2? 

Also, we should add a warn log maybe here? since we are skipping backpressure

Reply:
A lag of 1 is expected as the local refresh has happened and we are in the process of syncing the segment files. This has been kept as 2 to 1) allow writes to happen as due to segment merge the different lags (like bytes, time, etc) can jump instantly to a very high number and 2) allow one failure of upload to not kick in the back pressure too early.